src/BloomExe/web/EnhancedImageServer.cs

```diff
@@ -1,51 +1,137 @@
-﻿// Copyright (c) 2014 SIL International
+﻿// Copyright (c) 2014-2015 SIL International
 // This software is licensed under the MIT License (http://opensource.org/licenses/MIT)
 using System;
 using System.Diagnostics;
-using System.Drawing.Text;
 using System.Globalization;
-using System.Linq;
-using System.Net;
-using Bloom.ImageProcessing;
 using System.IO;
-using L10NSharp;
+using System.Threading;
 using Microsoft.Win32;
+using L10NSharp;
 using Palaso.IO;
 using Bloom.Collection;
-
+using Bloom.ImageProcessing;
 
 namespace Bloom.web
 {
 	/// <summary>
 	/// A local http server that can serve (low-res) images plus other files.
 	/// </summary>
+	/// <remarks>geckofx makes concurrent requests of URLs which this class handles. This means
+	/// that the methods of this class get called on different threads, so it has to be
+	/// thread-safe.</remarks>
 	public class EnhancedImageServer: ImageServer
 	{
 		private FileSystemWatcher _sampleTextsWatcher;
 		private bool _sampleTextsChanged = true;
+		private const int RWLockTimeOut = 100;
 
 		public CollectionSettings CurrentCollectionSettings { get; set; }
 
 		public EnhancedImageServer(LowResImageCache cache): base(cache)
 		{
+			Lock = new ReaderWriterLock();
+		}
+
+		// We use three different locks to synchronize access to the methods of this class.
+		// This allows certain methods to run concurrently.
+
+		// used to synchronize access to properties. Use this lock when you want to allow
+		// multiple concurrent readers, but no one else when writing the property.
+		private ReaderWriterLock Lock { get; set; }
+		// used to synchronize access to I18N methods
+		private object I18NLock = new object();
+		// used to synchronize access to various methods
+		private object SyncObj = new object();
+
+		private string _currentPageContent;
+		private string _accordionContent;
+		private bool _authorMode;
+
+		public string CurrentPageContent
```

Comment:
This is not a problem that needs fixing, but I'm curious how threading can cause a problem with a simple property like this. Even with locks, if several threads are reading and writing, it's a race which writes last and therefore wins, and a race which reader gets which of the values being written. If threads are reading and writing multiple properties, this much locking won't ensure that a consistent set of writes (all from the same thread) wins, or that no reader sees a state where one write has happened but another hasn't. What worse can happen without locks? It's not like an increment, where the read and write have to be indivisible.